### PR TITLE
Revert "make last work on any reversible collection (#42991), and fix it

### PR DESCRIPTION
**Pull request by vtjnash**
@MasonProtter @simeonschaub

### CODE REVIEW

File: base/iterators.jl [hunk 1]
```diff
@@ -882,6 +890,7 @@ function iterate(it::Cycle, state)
 end
 
 reverse(it::Cycle) = Cycle(reverse(it.xs))
+last(it::Cycle) = last(it.xs)
```
Comment (JeffBezanson) [inline base/iterators.jl:893]: This (and the method on the previous line!) doesn't seem right. The sequence doesn't converge.

**Comment by Sacha0** [comment]
Is anything preventing merging this revert? It seems worth unbreaking consumers' code till an alternative (e.g. https://github.com/JuliaLang/julia/pull/43110#issuecomment-971396451) lands? :)

**Comment by JeffBezanson** [comment]
It seems to me if we have the fallback chain last -> reverse -> indexing, then everything works for indexable objects, *plus* `last` continues to work for objects that only have `reverse`. For example, a doubly-linked list that you can reverse iterate efficiently, but not index efficiently.

**Comment by vtjnash** [comment]
Yes, but that was slightly more breaking still than necessary, and if you can decide to explicitly define the reverse iterator, then you can likely define this method easily also.

**Pull request merged**